src/Squidex.Infrastructure.MongoGridFs/Assets/MongoGridFsAssetStore.cs

```diff
@@ -0,0 +1,219 @@
+                {
+                    directory.Create();
+                }
+            }
+            catch (MongoException ex)
+            {
+                throw new ConfigurationException(
+                    $"Cannot connect to Mongo GridFS bucket '${bucket.Options.BucketName}'.", ex);
+            }
+            catch (IOException ex)
+            {
+                if (!directory.Exists)
+                {
+                    throw new ConfigurationException($"Cannot access directory '{directory.FullName}'", ex);
+                }
+            }
+        }
+
+        public string GenerateSourceUrl(string id, long version, string suffix)
+        {
+            var file = GetFile(id, version, suffix);
+
+            return file.FullName;
+        }
+
+        public async Task CopyAsync(string name, string id, long version, string suffix,
+            CancellationToken ct = default(CancellationToken))
+        {
+            try
+            {
+                var file = GetFile(name);
+
+                using (var stream = new MemoryStream())
+                {
+                    await bucket.DownloadToStreamByNameAsync(file.Name, stream, cancellationToken: ct);
+                    await bucket.UploadFromStreamAsync(file.Name, stream, cancellationToken: ct);
+                }
+
+                file.CopyTo(GetPath(id, version, suffix));
+            }
+            catch (FileNotFoundException ex)
+            {
+                throw new AssetNotFoundException($"Asset {name} not found.", ex);
+            }
+            catch (GridFSException ex)
+            {
+                throw new AssetNotFoundException($"Asset {name} not found.", ex);
+            }
+        }
+
+        public async Task DownloadAsync(string id, long version, string suffix, Stream stream,
+            CancellationToken ct = default(CancellationToken))
+        {
+            var file = GetFile(id, version, suffix);
+
+            try
+            {
+                if (file.Exists)
+                {
+                    using (var fileStream = file.OpenRead())
+                    {
+                        await fileStream.CopyToAsync(stream);
+                    }
+                }
+                else
+                {
+                    // file not found locally
+                    // read from GridFS
+                    await bucket.DownloadToStreamByNameAsync(file.Name, stream, cancellationToken: ct);
+
+                    // add to local assets
+                    using (var fileStream = file.OpenWrite())
+                    {
+                        await stream.CopyToAsync(fileStream);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                throw new AssetNotFoundException($"Asset {id}, {version} not found.", ex);
+            }
+        }
+
+        public Task UploadAsync(string name, Stream stream, CancellationToken ct = default(CancellationToken))
```

Comment:
Why? What are the cleanup rules? When do we delete them? What happens when you have multiple nodes? What is the advantage over simple folders?

GenerateSourceUrl is not that important. We can just return "UNSUPPORTED". The main reason is for debugging.

Reply:
If you want to keep a local copy I think it would make sense to make something like an "Aggregate" Asset store. Like a Multi Level Cache.

Reply:
expression bodies -> as you want Sir :)
GenerateSourceUrl  -> haven't digged too much into logic, if you say so then fine

Reply:
The real reason of the GenerateSourceUrl was to provide the url where you can download the file, e.g. when you make a public bucket you could the direct source url to bypass squidex. The implementation for the FileStorage does not really make sense.

Reply:
my idea was to keep local files and have GridFS as reserve store. so if i deploy application in a new machine it should be able to sync files from reserve place as GridFS. I would keep the possibility to serve static files here, bypassing squidex as you mentioned.

Reply:
sounds good

Reply:
May be IFileProvider abstraction would be better than IEnumerable. What you think? This would allow to store nested way too as bonus, if required of course. 

https://docs.microsoft.com/en-us/aspnet/core/fundamentals/file-providers?view=aspnetcore-2.1#file-provider-abstractions

Reply:
The IFileProvider is readonly, unfortunately. I mean something like this: https://github.com/aspnet/FileSystem/blob/dev/src/FS.Composite/CompositeFileProvider.cs

Reply:
CompositeFileProvider.cs just a set of IFileProviders. For reading fine